platform/linux-generic/include/odp_classification_internal.h

```diff
@@ -72,6 +77,97 @@ static inline void _odp_cos_queue_stats_add(cos_t *cos, odp_queue_t queue,
 		odp_atomic_add_u64(&cos->queue_stats[queue_idx].discards, discards);
 }
 
+/**
+ * Enqueue packets into destination CoS
+ */
+static inline void _odp_cos_enq(uint16_t cos_id, odp_queue_t dst, odp_packet_t packets[], int num)
+{
+	_ODP_ASSERT(cos_id != CLS_COS_IDX_NONE);
+	_ODP_ASSERT(dst != ODP_QUEUE_INVALID);
+
+	cos_t *cos = _odp_cos_entry_from_idx(cos_id);
+
+	if (num < 2 || !cos->vector.enable) {
+		int ret = odp_queue_enq_multi(dst, (odp_event_t *)packets, num);
+
+		if (odp_unlikely(ret != num)) {
+			if (ret < 0)
+				ret = 0;
+
+			odp_packet_free_multi(&packets[ret], num - ret);
+		}
+		_odp_cos_queue_stats_add(cos, dst, ret, num - ret);
+	} else {
+		_odp_cos_vector_enq(dst, (odp_event_t *)packets, num, cos);
+	}
+}
+
+/**
+ * Enqueue all remaining packets in 'packets' array
+ */
+static inline void _odp_cls_enq_all(odp_packet_t packets[], int num)
+{
+	odp_packet_hdr_t *prev_hdr;
+	odp_packet_hdr_t *latest_hdr = packet_hdr(packets[num - 1]);
+
+	if (num < 2) {
+		_odp_cos_enq(latest_hdr->cos, latest_hdr->dst_queue, packets, 1);
+		return;
+	}
+
+	prev_hdr = packet_hdr(packets[num - 2]);
+
+	if (prev_hdr->dst_queue == latest_hdr->dst_queue && prev_hdr->cos == latest_hdr->cos) {
+		_odp_cos_enq(prev_hdr->cos, prev_hdr->dst_queue, packets, num);
+	} else {
+		_odp_cos_enq(prev_hdr->cos, prev_hdr->dst_queue, packets, num - 1);
+		_odp_cos_enq(latest_hdr->cos, latest_hdr->dst_queue, &packets[num - 1], 1);
+	}
+}
+
+/**
+ * Enqueue packets into classifier destination queue
+ *
+ * Called by pktio devices for each received packet when classifier has been enabled. Postpones the
+ * actual enqueue operation and stores packets in 'packets' array until destination queue or CoS
+ * change, or 'last' flag is set.
+ *
+ * @param[out] packets Packet array to be enqueued
+ * @param      num     Number of handles in 'packets' array
+ * @param      last    Enqueue all packets
+ *
+ * @return Number of packets remaining in 'packets' array
+ */
+static inline int _odp_cls_enq(odp_packet_t packets[], int num, odp_bool_t last)
```

Comment:
I wonder if the code could be simplified if some state was kept in a helper struct. Probably the packet array, number of packets stored, current queue/cos, the total number of packets to be processed would be needed. Drivers would pass the struct, a packet and current index. The "last" step could then be done without a separate function call and flag. Any sense?

Reply:
The main motivation for the current solution was to minimize the amount of changes required to the pktio device code, since the integration had to be done into seven different drivers. I'd prefer not changing `_odp_cls_enq()` function at this point, unless this is deemed necessary. Mainly because this would be a rather big change and so it would invalidate most of the testing done so far. However, this could be added as a classifier optimization task to the backlog.

Reply:
OK, perhaps a future optimization then.